File: py/aoc2021/day6.py
def solve(times, lines):
    fishes = [0] * 9
    for line in lines:
        for fish in line.strip().split(","):
            fishes[int(fish)] += 1
    for _ in range(times):
        zero = fishes[0]
        fishes[0:8] = fishes[1:]
        fishes[6] += zero
        fishes[8] = zero
    return sum(fishes)


def part1(lines):
    """
    >>> part1(["3,4,3,1,2"])
    5934
    """
    return solve(80, lines)


def part2(lines):
    """
    >>> part2(["3,4,3,1,2"])
    26984457539
    """
    return solve(256, lines)


parts = (part1, part2)

File: rs/src/day6.rs
use super::util;
use std::error::Error;

build_const!("libaoc2021_day6");

fn solve<'a, I, S, const N: usize>(
    lut: &[usize; N],
    lines: I,
) -> Result<usize, Box<dyn Error + Send + Sync>>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    let mut ret = 0;
    for line in lines {
        for word in line.as_ref().split(',') {
            ret += lut.get(word.parse::<usize>()?).ok_or(util::Error)?;
        }
    }
    Ok(ret)
}

pub fn part1<'a, I, S>(lines: I) -> Result<usize, Box<dyn Error + Send + Sync>>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    solve(&LUT_80, lines)
}

pub fn part2<'a, I, S>(lines: I) -> Result<usize, Box<dyn Error + Send + Sync>>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    solve(&LUT_256, lines)
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    static EXAMPLE: &[&str] = &["3,4,3,1,2"];

    #[test]
    fn part1_examples() -> Result<(), Box<dyn Error + Send + Sync>> {
        assert_eq!(5934, part1(EXAMPLE)?);
        Ok(())
    }

    #[test]
    fn part2_examples() -> Result<(), Box<dyn Error + Send + Sync>> {
        assert_eq!(26984457539, part2(EXAMPLE)?);
        Ok(())
    }
}
